### PR TITLE
Replace some usages of fmt.Sprintf with more efficient string concatenation

### PR DESCRIPTION
**Pull request by schlosna**
As mentioned in https://github.com/cilium/cilium/issues/19571 `fmt.Sprintf()` is well-known for not being opimal in terms of CPU and memory allocations. This PR replaces some usages of `fmt.Sprintf` on possible hot code paths with more efficient string concatenation.

* [pkg/policy] Benchmark ProxyID
* [pkg/policy] ProxyID uses string concat instead of Sprintf
* [pkg/policy] L4PolicyMap uses string concatenation
* [pkg/policy] map state uses string concatenation
* [pkg/policy] rulePortsCoverSearchContext uses string concatenation
* [pkg/policy] NewVisibilityPolicy uses string concatenation
* [pkg/policy] groups use string concatenation
* [pkg/loadbalancer] Benchmark L3n4Addr String and StringWithProtocol
* [pkg/loadbalancer] L3n4Addr String and StringWithProtocol use string concatenation
* [pkg/kvstore] makeSessionName and RunLocksGC use string concatenation
* [pkg/endpoint] Use string concatenation
* [operator/pkg/model] Ingress listener name uses string concatenation

Related: https://github.com/cilium/cilium/issues/19571

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [ ] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

```release-note
Replace some usages of fmt.Sprintf with more efficient string concatenation
```

### CODE REVIEW

**Comment by schlosna** [comment]
`benchstat` output, old is (main) ec9cb2393601615c5db211ced1e558e4cce6790b + 2 benchmark commits (https://github.com/cilium/cilium/pull/27518/commits/8d188a5e6ce956f98f61fdae8647ae700ad69d8a + https://github.com/cilium/cilium/pull/27518/commits/5ddd7a41662de8947bf9b861303fd3e6e98f84e3 )

```
goos: darwin
goarch: arm64
pkg: github.com/cilium/cilium/pkg/loadbalancer
                                        │ old-loadbalancer.txt │         new-loadbalancer.txt         │
                                        │        sec/op        │    sec/op     vs base                │
L3n4Addr_Hash_IPv4-10                             21.63n ±  1%   22.10n ± 13%   +2.17% (p=0.001 n=10)
L3n4Addr_Hash_IPv6_Short-10                       21.63n ±  1%   21.95n ±  3%   +1.48% (p=0.007 n=10)
L3n4Addr_Hash_IPv6_Long-10                        21.81n ±  2%   22.00n ±  3%        ~ (p=0.210 n=10)
L3n4Addr_Hash_IPv6_Max-10                         22.03n ± 11%   21.97n ± 10%        ~ (p=0.986 n=10)
L3n4Addr_String_IPv4-10                          150.65n ±  1%   74.55n ± 13%  -50.52% (p=0.000 n=10)
L3n4Addr_String_IPv6_Max-10                       183.3n ± 15%   121.7n ± 26%  -33.62% (p=0.000 n=10)
L3n4Addr_StringWithProtocol_IPv4-10              186.65n ± 11%   79.98n ±  5%  -57.15% (p=0.000 n=10)
L3n4Addr_StringWithProtocol_IPv6_Max-10           213.3n ±  4%   125.4n ± 14%  -41.23% (p=0.000 n=10)
geomean                                           62.97n         46.36n        -26.38%

                                        │ old-loadbalancer.txt │         new-loadbalancer.txt         │
                                        │         B/op         │    B/op     vs base                  │
L3n4Addr_Hash_IPv4-10                               24.00 ± 0%   24.00 ± 0%        ~ (p=1.000 n=10) ¹
L3n4Addr_Hash_IPv6_Short-10                         24.00 ± 0%   24.00 ± 0%        ~ (p=1.000 n=10) ¹
L3n4Addr_Hash_IPv6_Long-10                          24.00 ± 0%   24.00 ± 0%        ~ (p=1.000 n=10) ¹
L3n4Addr_Hash_IPv6_Max-10                           24.00 ± 0%   24.00 ± 0%        ~ (p=1.000 n=10) ¹
L3n4Addr_String_IPv4-10                             74.00 ± 0%   44.00 ± 0%  -40.54% (p=0.000 n=10)
L3n4Addr_String_IPv6_Max-10                         114.0 ± 0%   101.0 ± 0%  -11.40% (p=0.000 n=10)
L3n4Addr_StringWithProtocol_IPv4-10                 98.00 ± 0%   52.00 ± 0%  -46.94% (p=0.000 n=10)
L3n4Addr_StringWithProtocol_IPv6_Max-10             146.0 ± 0%   117.0 ± 0%  -19.86% (p=0.000 n=10)
geomean                                             50.16        41.60       -17.06%
¹ all samples are equal

                                        │ old-loadbalancer.txt │         new-loadbalancer.txt         │
                                        │      allocs/op       │ allocs/op   vs base                  │
L3n4Addr_Hash_IPv4-10                               1.000 ± 0%   1.000 ± 0%        ~ (p=1.000 n=10) ¹
L3n4Addr_Hash_IPv6_Short-10                         1.000 ± 0%   1.000 ± 0%        ~ (p=1.000 n=10) ¹
L3n4Addr_Hash_IPv6_Long-10                          1.000 ± 0%   1.000 ± 0%        ~ (p=1.000 n=10) ¹
L3n4Addr_Hash_IPv6_Max-10                           1.000 ± 0%   1.000 ± 0%        ~ (p=1.000 n=10) ¹
L3n4Addr_String_IPv4-10                             5.000 ± 0%   3.000 ± 0%  -40.00% (p=0.000 n=10)
L3n4Addr_String_IPv6_Max-10                         4.000 ± 0%   3.000 ± 0%  -25.00% (p=0.000 n=10)
L3n4Addr_StringWithProtocol_IPv4-10                 6.000 ± 0%   3.000 ± 0%  -50.00% (p=0.000 n=10)
L3n4Addr_StringWithProtocol_IPv6_Max-10             5.000 ± 0%   3.000 ± 0%  -40.00% (p=0.000 n=10)
geomean                                             2.225        1.732       -22.14%
¹ all samples are equal
```

File: pkg/policy/l4.go [hunk 1]
```diff
@@ -1138,32 +1138,44 @@ func (l4 L4PolicyMap) containsAllL3L4(labels labels.LabelArray, ports []*models.
...
 		switch lwrProtocol {
 		case "", models.PortProtocolANY:
-			tcpPort := fmt.Sprintf("%s/TCP", portStr)
+			tcpPort := portStr + "/TCP"
 			tcpFilter, tcpmatch := l4[tcpPort]
 			if tcpmatch {
 				tcpmatch, isTCPDeny = tcpFilter.matchesLabels(labels)
+				if isTCPDeny {
+					return api.Denied
+				}
```
Comment (christarazi) [inline pkg/policy/l4.go]: This change is a subtle change in behavior. I'd prefer if it was left out of this PR and done in a separate one, in the case it somehow causes a regression.

**Review by christarazi — CHANGES_REQUESTED** [review]
Thanks for the PR! Just one comment.

**Comment by joestringer** [comment]
It looks like this subtly changed LB frontend string formatting which broke a bunch of golden test cases: https://app.travis-ci.com/github/cilium/cilium/builds/265504327 .

I've also kicked off the full CI to get feedback from the rest of the testsuite. Those results should come in over the next 30-60 minutes.

File: pkg/policy/l4.go (hunk 1 above)
Comment (schlosna) [reply pkg/policy/l4.go]: makes sense, I'll revert this out of this PR and submit separately if desired

**Comment by schlosna** [comment]
> It looks like this subtly changed LB frontend string formatting which broke a bunch of golden test cases: https://app.travis-ci.com/github/cilium/cilium/builds/265504327 .

Sorry about those failures. I'm adding some unit tests for LB `String()` and `StringWithProtocol()` and fixing these breaks.

**Review by meyskens — APPROVED** [review]
LGTM for service mesh changes

**Review by giorio94 — APPROVED** [review]
/lgtm for kvstore changes.

**Comment by schlosna** [comment]
@christarazi I see several `Conformance` tests failed. I'm not very familiar with cilium CI yet, so not sure how to interpret these. From looking through the logs it appears to me that there are some connection failures. What is the the best way to debug and iterate on these types of CI failures?

**Comment by christarazi** [comment]
* Runtime hit https://github.com/cilium/cilium/issues/27761
* EKS failed to provision cluster
* E2E hit https://github.com/cilium/cilium/issues/27762

**Comment by christarazi** [comment]
The failures all seemed to be flakes. I've re-run them.

**Review by christarazi — APPROVED** [review]

**Pull request merged**